### PR TITLE
add diagnostics for E0387

### PR DESCRIPTION
**Pull request by AlisdairO**
As title :-)
Part of #24407.

r? @Manishearth

### CODE REVIEW

**Comment by Manishearth** [comment]
Welcome back!

**Comment by Manishearth** [comment]
I'd like to see some tips for cases where the function definition is not owned by the user too.

Something that hints the usage of interior mutability (std::cell) would be nice.

**Comment by AlisdairO** [comment]
Thanks!  I'm not exactly firing on all cylinders, but hoping to get things done here and there :-).

Good point on the use of cell, that actually hadn't crossed my mind.  I'll add it in.

**Comment by AlisdairO** [comment]
Added info about cell types!

File: src/librustc_borrowck/diagnostics.rs [hunk 1]
```diff
@@ -95,6 +95,58 @@ fn main(){
...
+fn foo<F: FnMut()>(f: F) { }
+```
+
+Alternatively, we can consider using the `Cell` and `RefCell` types to achieve
+interior mutability through a shared reference. Our example's `mutable` function
+could be redefined as below:
+
+```
+fn mutable() {
+    let x = std::cell::Cell::new(0u32);
```
Comment (Manishearth) [inline src/librustc_borrowck/diagnostics.rs]: nit: just use `import std::cell::Cell;` at the top, long paths look ugly inline

Comment (AlisdairO) [inline src/librustc_borrowck/diagnostics.rs]: done :)

**Pull request merged**